lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -823,6 +846,90 @@ data ByronWalletPutPassphraseData = ByronWalletPutPassphraseData
     , newPassphrase :: !(ApiT (Passphrase "raw"))
     } deriving (Eq, Generic, Show)
 
+-- | Polymorphic wrapper for byte arrays, parameterised by the desired string
+-- encoding.
+newtype ApiBytesT (base :: Base) bs = ApiBytesT { getApiBytesT :: bs }
+    deriving (Generic, Show, Eq, Functor)
+    deriving newtype (Semigroup, Monoid, Hashable)
+    deriving anyclass NFData
+
+newtype ApiSerialisedTransaction
+    = ApiSerialisedTransaction (ApiBytesT 'Base64 SerialisedTx)
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+data ApiConstructTransaction (n :: NetworkDiscriminant) = ApiConstructTransaction
+    { serializedTransaction :: !ApiSerialisedTransaction
+    , coinSelection :: !(ApiCoinSelection n)
+    , fee :: !(Quantity "lovelace" Natural)
+    } deriving (Eq, Generic, Show)
+      deriving anyclass NFData
+
+-- | Index of the stake key.
+newtype StakeKeyIndex = StakeKeyIndex (ApiT DerivationIndex)
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+-- | Stake pool delegation certificates.
+data ApiMultiDelegationAction
+    = Joining !(ApiT PoolId) !StakeKeyIndex
+    -- ^ Delegate given staking index to a pool, possibly registering the stake
+    -- key at the same time.
+    | Leaving !StakeKeyIndex
+    -- ^ Undelegate the given staking index from its pool.
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+-- | Input parameters for transaction construction.
+data ApiConstructTransactionData (n :: NetworkDiscriminant) = ApiConstructTransactionData
+    { payments :: !(Maybe (ApiPaymentDestination n))
+    , withdrawal :: !(Maybe ApiWithdrawalPostData)
+    , metadata :: !(Maybe (ApiT TxMetadata))
+    , mint :: !(Maybe (ApiT W.TokenMap))
+    , delegations :: !(Maybe [ApiMultiDelegationAction])
+    , validityInterval :: !(Maybe ApiValidityInterval)
+    , passphrase :: !(ApiT (Passphrase "lenient"))
+    } deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+{--
+-- | New transaction API version of 'AddressAmount'.
+data AddressValue addr = AddressValue
+    { address :: !addr
+    , value :: !(ApiT W.TokenBundle)
+    } deriving (Eq, Generic, Show)
+      deriving anyclass NFData
+--}
+
+data ApiPaymentDestination (n :: NetworkDiscriminant)
+    = ApiPaymentAddresses !(NonEmpty (AddressAmount (ApiAddressIdT n)))
```

Comment:
If using `NonEmpty` for `ApiPaymentAddresses`, then I suggest adding back the "None" constructor here, and removing the `Maybe` from `ApiConstructTransactionData`.
```suggestion
    = ApiPaymentNone
    --  ^ The only output is a change address.
    | ApiPaymentAddresses !(NonEmpty (AddressAmount (ApiAddressIdT n)))
```

Reply:
well, we have `Maybe ApiPaymentDestination` used so optionality is captured in that way. I am slightly in favor of what it is now, but will not argue a lot;-)

Reply:
Yeah actually it seems like the JSON instances are a bit easier when using Maybe.